### PR TITLE
feat(githubpolicy): snapshot v3 — group layer and endpoint directory identity

### PR DESCRIPTION
**Pull request by michiosw**
**C2 of 2** — SCIM group policies, CLI side (ENG-463; server stack: kontext #618–#622, needs #621/K4 deployed for v3 responses).

- Snapshot negotiation: requests `?schema=github-policy-snapshot-v3&installation_id=ins_…`, accepts v2 **or** v3 for one release (a pre-K4 server ignores the params and answers v2). A follow-up release pins v3-only.
- New `group` layer: matches by set membership against `endpointDirectory.groupIds` (the endpoint's SCIM directory identity, resolved server-side from the C1-reported email). Missing/unmatched/ambiguous identity → empty set → group rules inert, fail closed.
- Layer rank renumbered `org(0) < group(1) < user/agent/endpoint(2)` — dimension count still dominates, exact ties still deny; the entire pre-existing evaluator test table passes unchanged (regression net).
- Fail-closed guards: group rule in a v2 body → reject (server contract filters them); unknown layer → reject.
- Dry-run audit context gains `groups_resolved` and reports the actual evaluated `schema_version`.

Evaluator matrix mirrors the K4 server specs verbatim (org vs group, group vs endpoint, multi-group tie → deny, dimensions-before-layer).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by michiosw** [comment]
* **#347** <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/347?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/347?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#346** <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/346?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: internal/githubpolicy/client.go [hunk 1]
```diff
@@ -19,8 +19,13 @@ const maxSnapshotBodyBytes = 4 << 20
...
 // same per-customer install token as the authorization-ledger ingest.
-func FetchSnapshot(ctx context.Context, client *http.Client, cloudURL, installToken string) (Snapshot, error) {
-	endpoint, err := snapshotURL(cloudURL)
+//
+// installationID identifies this endpoint ("ins_…") so the cloud can resolve
+// its directory identity for group-layer rules; empty is allowed and simply
+// yields no group matches. The request asks for schema v3 but accepts v2 for
+// one release: a pre-v3 server ignores the query params and answers v2.
+func FetchSnapshot(ctx context.Context, client *http.Client, cloudURL, installToken, installationID string) (Snapshot, error) {
+	endpoint, err := snapshotURL(cloudURL, installationID)
```
[the remark this answers is not in the document]
Comment (michiosw) [reply internal/githubpolicy/client.go:31]: Fixed in the latest push: installationID is now trimmed once at the FetchSnapshot entry point, so the query param and the echoed-directory check see the same value.

File: internal/githubpolicy/evaluate.go [hunk 2]
```diff
@@ -55,6 +55,14 @@ type Evaluation struct {
...
 	// available, so user/agent-layer rules could not match their subject.
 	SubjectsResolved bool
+	// GroupsResolved is false when the snapshot carried no directory group
+	// ids for this endpoint (v2 snapshot, no installation id sent, or
+	// missing/unmatched/ambiguous email), so group-layer rules could not
+	// match.
+	GroupsResolved bool
+	// SchemaVersion is the wire format of the snapshot that produced this
+	// evaluation (v2 or v3 during the negotiation window).
+	SchemaVersion string
```
[the remark this answers is not in the document]
Comment (michiosw) [reply internal/githubpolicy/evaluate.go]: Fixed in the latest push: GroupsResolved now means "directory identity resolved to a user" (DirectoryUserID != nil for this endpoint), so a matched user with zero group memberships reports true; the field comment enumerates the false cases.

File: internal/githubpolicy/snapshot.go [hunk 3]
```diff
@@ -88,13 +103,33 @@ type Snapshot struct {
...
+// (ambiguity fails closed — never a union of candidate groups); GroupIDs is
+// then empty and group-layer rules can never match.
+type EndpointDirectory struct {
+	InstallationID  string   `json:"installationId"`
+	DirectoryUserID *string  `json:"directoryUserId"`
+	GroupIDs        []string `json:"groupIds"`
+}
+
+
 // Enforce reports whether the snapshot explicitly directs enforcement.
```
[the remark this answers is not in the document]
Comment (michiosw) [reply internal/githubpolicy/snapshot.go:132]: Fixed in the latest push (gofmt run over the package).

**Comment by tumberger** [comment]
Detailed review of the stack (Claude-assisted). C2 came back clean — nits only:

**1. [Nit] Redundant `TrimSpace` on `installationID`** — trimmed in `FetchSnapshot` (`client.go:30`) and again inside `snapshotURL` (`client.go:129`); the second reassigns the parameter to no effect.

**2. [Nit] Outer `cacheFile.SchemaVersion` is now vestigial** (`cache.go`) — it's written from `snapshot.SchemaVersion` but `LoadPersisted` only checks the inner `file.Snapshot.SchemaVersion`. A comment noting it's kept for cache-file debugging/back-compat would prevent someone treating it as load-bearing.

**3. [Nit] Missing edge-case test** — no `FetchSnapshot` test for `installationID=""` + v3 response; the code intentionally accepts a v3 body without a directory block in that case, and a test would document that acceptance.

Verified during review (the important claims hold): version detection is by explicit `schemaVersion` echo, never heuristics — no misclassification path; all reject paths (group-rule-in-v2, unknown layer, v3-without-directory, foreign directory) fall back to the stale cached snapshot with diagnostic logging, and a fresh start with a bad body skips evaluation rather than misapplying; the layer renumbering is safe for cached v2 snapshots (only relative rank order matters in `rulePrecedes`, and v2 caches contain no group rules by construction); the pre-existing evaluator table passes unchanged except the constant rename; and the new group matrix mirrors the K4 server specs including the two-group exact-tie → deny case.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by tumberger — DISMISSED** [review]
